url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/plugins/xml/XMLSchedulingDataProcessor.java

```diff
@@ -0,0 +1,592 @@
+      logger.warn("Could not load jobs schema from classpath!");
+    } else {
+      inputSource = new InputSource(is);
+    }
+
+    return inputSource;
+  }
+
+  /**
+   * Process the xml file in the given location, and schedule all of the jobs defined within it.
+   *
+   * @param fileName meta data file name.
+   */
+  public void processFile(String fileName, boolean failOnFileNotFound) throws Exception {
+
+    boolean fileFound = false;
+    InputStream f = null;
+    try {
+      String furl = null;
+
+      File file = new File(fileName); // files in filesystem
+      if (!file.exists()) {
+        URL url = classLoadHelper.getResource(fileName);
+        if (url != null) {
+          try {
+            furl = URLDecoder.decode(url.getPath(), "UTF-8");
+          } catch (UnsupportedEncodingException e) {
+            furl = url.getPath();
+          }
+          file = new File(furl);
+          try {
+            f = url.openStream();
+          } catch (IOException ignor) {
+            // Swallow the exception
+          }
+        }
+      } else {
+        try {
+          f = new java.io.FileInputStream(file);
+        } catch (FileNotFoundException e) {
+          // ignore
+        }
+      }
+
+      if (f == null) {
+        fileFound = false;
+      } else {
+        fileFound = true;
+      }
+    } finally {
+      try {
+        if (f != null) {
+          f.close();
+        }
+      } catch (IOException ioe) {
+        logger.warn("Error closing jobs file " + fileName, ioe);
+      }
+    }
+
+    if (!fileFound) {
+      if (failOnFileNotFound) {
+        throw new SchedulerException("File named '" + fileName + "' does not exist.");
+      } else {
+        logger.warn("File named '" + fileName + "' does not exist. This is OK if you don't want to use an XML job config file.");
+      }
+    } else {
+      processFile(fileName);
+    }
+  }
+
+  /**
+   * Process the xmlfile named <code>fileName</code> with the given system ID.
+   *
+   * @param fileName meta data file name.
+   * @param systemId system ID.
+   */
+  private void processFile(String fileName) throws ValidationException, ParserConfigurationException, SAXException, IOException, SchedulerException,
+      ClassNotFoundException, ParseException, XPathException {
+
+    prepForProcessing();
+
+    logger.info("Parsing XML file: " + fileName);
+    InputSource is = new InputSource(getInputStream(fileName));
+    // is.setSystemId(systemId);
+
+    process(is);
+
+    maybeThrowValidationException();
+  }
+
+  /*
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Interface.
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   */
+
+  private void prepForProcessing() {
+
+    clearValidationExceptions();
+
+    loadedJobs.clear();
+    loadedTriggers.clear();
+  }
+
+  private void process(InputSource is) throws SAXException, IOException, ParseException, XPathException, ClassNotFoundException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 29. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.